solidity/contracts/converter/BancorConverter.sol

```diff
@@ -646,22 +666,21 @@ contract BancorConverter is IBancorConverter, SmartTokenController, ContractRegi
         uint256 toReserveBalance = getReserveBalance(_toToken);
         assert(amount < toReserveBalance);
 
-        // transfer funds from the caller in the from reserve token
-        if (_fromToken == IERC20Token(0)) {
+        // ensure that the input amount was already deposited
+        if (_fromToken == IERC20Token(0))
             require(msg.value == _amount);
-        }
-        else {
-            require(msg.value == 0);
-            ensureTransferFrom(_fromToken, msg.sender, this, _amount);
-        }
+        else
+            require(msg.value == 0 && _fromToken.balanceOf(this).sub(getReserveBalance(_fromToken)) >= _amount);
 
-        // transfer funds to the caller in the to reserve token
-        if (_toToken == IERC20Token(0)) {
-            msg.sender.transfer(amount);
-        }
-        else {
-            ensureTransferFrom(_toToken, this, msg.sender, amount);
-        }
+        // sync the reserve balances
+        syncReserveBalance(_fromToken);

```

Comment:
May as well embed the code of function `syncReserveBalance` into the if/else clause above.